### ISSUE TITLE
Tensorflow only using a fraction of the GPU power available (NLP model, dual GPU, tf.data pipeline) 

### ISSUE THREAD

**Issue by philipp-ghtest** [tensorflow/tensorflow#42820]
Hi!

Somewhat recently I got a new training server which is really fast, but I'm currently having trouble utilizing it's GPU and CPU to it's full potential when training my model.

I'm training an NLP classification model with a string as input and a category as target. When I set the batch size to a reasonally small number, like 16 or 32, only around 10% of each of the 2 GPUs as well as of the CPU are used. Only when I size the batches up to 4096, CPU gets close to a 100% load but GPUs still only hit 7-8%. Training is really fast then but extremely inefficient because such batch sizes are b\*\*\*s\*\*\*, so the model converges only very slowly.

I found a sweet spot around bs=256, where only around 20% CPU and 10% GPUs load is achieved and gradient descent is still somewhat efficient, which means I get the best results in terms of wall time.

The data pipeline is implemented with [tf.data](https://tf.data), reading the data from several CSVs in parallel from an SSD. I couldn't find any bottlenecks so far.

This is somewhat frustrating because I can only make use of a fraction of the full potential of my new machine. Any ideas on how to improve this?

I'm grateful for any help.

&#x200B;

&#x200B;

My specs:

\- AMD Ryzen Threadripper 3960X 24-Core Processor

\- 64 GB RAM

\- two NVIDIA GeForce RTX 2070 SUPER with 8192MiB each

\- Win 10 (unfortunately, the ASrock Creator TRX40 motherboard we bought is currently incompatible with Linux, wtf...)

\- TF 2.1.0 installed from binary (anaconda)

\- Python 3.7.7

\- CUDA Version 10.2.89

&#x200B;

The relevant part of my code:

    class dataset_loader():
        
        def __init__(self, data_dir, csv_file, batch_size, cycle_length, tokenizer=None, n_threads=1, n_prefetch=1):
            
            self.batch_size = batch_size
            self.cycle_length = cycle_length
            self.n_threads = n_threads
            self.n_prefetch = n_prefetch
            
            self.output_size = 3943     ## !!!!!!!!!!! TODO nur für testzwecke bitte nicht hardcoden!!!!!!!!!!!
    
            if(tokenizer is None):
                self.tokenizer = tf.keras.preprocessing.text.Tokenizer(char_level=True)
                self.tokenizer.fit_on_texts(strat_search_words_with_beginnings)
            else:
                self.tokenizer = tokenizer
                
            char_dict = list(eval(self.tokenizer.get_config().get("word_index")).keys())[:-1]  # hier vorletztes weglassen, da ein Out-Of-Vocabulary slot bei StaticVocabularyTable zwingend angegeben werden muss
            char_index = list(eval(self.tokenizer.get_config().get("word_index")).values())[:-1]
            char_table_init = tf.lookup.KeyValueTensorInitializer(char_dict, char_index, value_dtype=tf.int64)
            self.char_table = tf.lookup.StaticVocabularyTable(char_table_init, 1)
            
            self.max_id = len(self.tokenizer.word_index)
            
            assert self.max_id == len(char_index)+1
    
            self._filepath_ = os.path.join(data_dir, csv_file)
    
        
        @tf.function
        def preprocessing(self, line):
            split_line = tf.strings.split(line, sep=";")
            
            search_string = split_line[0]
            search_string = tf.strings.substr(search_string, 0, 50)      # zu lange strings abschneiden
            while tf.strings.length(search_string) < 50:                 # zu kurze strings padden
                search_string = search_string + tf.constant(" ")
    
            chars = tf.strings.bytes_split(search_string)
            indices = self.char_table.lookup(chars)
            one_hot_string = tf.one_hot(indices, depth=self.max_id+1)
            
            icd_idx = tf.strings.to_number(split_line[1], out_type=tf.dtypes.int32)
            one_hot_icd = tf.one_hot(icd_idx, self.output_size)
            
            return one_hot_string, one_hot_icd
        
        def interleaving(self, filepath):
            return tf.data.TextLineDataset(filepath).skip(1)
        
        def get_dataset(self):
            dataset = tf.data.Dataset.list_files(self._filepath_)
            dataset = dataset.interleave(self.interleaving, cycle_length=self.cycle_length, num_parallel_calls=self.n_threads)
            dataset = dataset.map(self.preprocessing, num_parallel_calls=self.n_threads)
            return dataset.batch(self.batch_size).prefetch(self.n_prefetch)
        
        def get_tokenizer(self):
            return self.tokenizer
           
    
    
    import pickle
    f = open(os.path.join(data_dir, "tokenizer.pkl"), "rb")
    tokenizer = pickle.load(f)
    
    train_data_loader = dataset_loader(data_dir,
                                 "train_*.csv",
                                 batch_size=batch_size,
                                 tokenizer=tokenizer,
                                 cycle_length=106,             #Anzahl der CSV-Dateien
                                 n_threads=tf.data.experimental.AUTOTUNE,
                                 n_prefetch=40)
    
    valid_data_loader = dataset_loader(data_dir,
                                "valid_*.csv",
                                batch_size=batch_size,
                                tokenizer=tokenizer,
                                cycle_length=13,
                                n_threads=tf.data.experimental.AUTOTUNE,
                                n_prefetch=40)
    
    
    mirrored_strategy = tf.distribute.MirroredStrategy(cross_device_ops=tf.distribute.HierarchicalCopyAllReduce())
    
    with mirrored_strategy.scope():
        model = keras.models.Sequential([
        #    keras.layers.GRU(128, return_sequences=True, batch_input_shape=[batch_size, None, max_id+1]),
            keras.layers.GRU(128, return_sequences=True, input_shape=[ None, train_data_loader.max_id+1], use_bias=False),
            keras.layers.GRU(128, return_sequences=True, use_bias=False),
            keras.layers.GRU(128, use_bias=False),
            keras.layers.Flatten(),
            keras.layers.Dense(train_data_loader.output_size, activation="softmax")
        ])
        model.compile(loss=[focal_loss_umbertogriffo.categorical_focal_loss(alpha=.25, gamma=2)], optimizer="adam", metrics=['accuracy'])
    
    callbacks = list()
    callbacks.append(keras.callbacks.EarlyStopping(patience=2))
    callbacks.append(keras.callbacks.ModelCheckpoint(filepath = os.path.join(data_dir, "checkpoints"), save_best_only=True))
    
    history = model.fit(train_data_loader.get_dataset(), validation_data=valid_data_loader.get_dataset(), epochs=25, callbacks = callbacks)

**Comment by jvishnuvardhan** [comment]
@philipp-ghtest We donot support Anaconda builds as we don't have complete knowledge on how they were built. May be you need to post it in Anaconda repository.

Generally, pip builds for 2.1 are built with CUDA 10.1 (you mentioned you have CUDA 10.2).Please check [here](https://www.tensorflow.org/install/source_windows#gpu) for more details.


Version | Python version | Compiler | Build tools | cuDNN | CUDA
-- | -- | -- | -- | -- | --
tensorflow_gpu-2.3.0 | 3.5-3.8 | MSVC 2019 | Bazel 3.1.0 | 7.4 | 10.1
tensorflow_gpu-2.2.0 | 3.5-3.8 | MSVC 2019 | Bazel 2.0.0 | 7.4 | 10.1
tensorflow_gpu-2.1.0 | 3.5-3.7 | MSVC 2019 | Bazel 0.27.1-0.29.1 | 7.4 | 10.1

Two options are 
(1) post it in Anaconda's repo where experts related to Anaconda will resolve your issue, or
(2) unistall TF, CUDA drivers, clean up any remaining files related to CUDA/TF, restart, then reinstall pip version freshly.

Hope it helps. Please let us know how it progresses. Thanks!
